notebooks/ai-ml/SparkXGBoostCustomerChurn.ipynb

```diff
@@ -0,0 +1,838 @@
+    "print(f\"Accuracy: {accuracy:.4f}\")\n",
+    "\n",
+    "f1 = evaluator_multi.evaluate(predictions, {evaluator_multi.metricName: \"f1\"})\n",
+    "print(f\"F1 Score: {f1:.4f}\")\n",
+    "\n",
+    "weighted_precision = evaluator_multi.evaluate(predictions, {evaluator_multi.metricName: \"weightedPrecision\"})\n",
+    "print(f\"Weighted Precision: {weighted_precision:.4f}\")\n",
+    "\n",
+    "weighted_recall = evaluator_multi.evaluate(predictions, {evaluator_multi.metricName: \"weightedRecall\"})\n",
+    "print(f\"Weighted Recall: {weighted_recall:.4f}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c3735271-a6b3-4077-b80e-c362eb22cacf",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Save to Cloud storage\n",
+    "\n",
+    "Save the train data, test data, and XGBoost model to Cloud storage."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "86d3964a-4314-4074-a673-efd1d7016e36",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Specify GCS bucket training data path\n",
+    "data_path = f\"gs://{bucket}/curated/models/churn_prediction\"\n",
+    "\n",
+    "# Save transformed training data\n",
+    "train_data.write.parquet(f\"{data_path}/train_data\", mode=\"overwrite\")\n",
+    "\n",
+    "# Save transformed test data\n",
+    "test_data.write.parquet(f\"{data_path}/test_data\", mode=\"overwrite\")\n",
+    "\n",
+    "# Save the trained model to GCS:\n",
+    "model.write().overwrite().save(f\"{data_path}/xgboost_model\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cceb3128-cef5-496b-ba94-bc3c3c7622c0",
+   "metadata": {},
+   "source": [
+    "# Run Batch Inference on Spark\n",
+    "\n",
+    "Load the model and test data from Cloud Storage into Spark."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "41e4f56d-494d-434a-8540-07d720582a9b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Specify my GCP bucket\n",
+    "model_bucket = f\"gs://{BUCKET}/curated/models/churn_prediction\"\n",
+    "\n",
+    "# Load transformed training data\n",
+    "test_data_loaded = spark.read.parquet(f\"{model_bucket}/test_data\")\n",
+    "\n",
+    "# Load the trained model\n",
+    "model_loaded = PipelineModel.load(f\"{model_bucket}/xgboost_model\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "828831e8-e043-459f-a68a-19e5e0377151",
+   "metadata": {},
+   "source": [
+    "Generate predictions on the test data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b28c16db-1b66-460a-aafb-2762d23d0062",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "predictions = model_loaded.transform(test_data_loaded)\n",
+    "\n",
+    "# Filter down to only include the customer_id and prediction columns\n",
+    "predictions = predictions.select(\"customer_id\", \"prediction\")\n",
```

Comment:
Consider including a few more columns as such:

`predictions = predictions.select("customer_id", "prediction", "rawPrediction", "probability")` 

Though not totally necessary, having that additional context in BQ could be helpful.